#! /bin/sh /usr/share/dpatch/dpatch-run
## 00_0167-fix-group-reading.dpatch by Dirk Mueller <mueller@kde.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: In big user environments, getgrgid_r() needs more memory than sysconf() returns.

@DPATCH@
diff -urNad qt4-x11-4.3.0~rc1~/src/corelib/io/qfsfileengine_unix.cpp qt4-x11-4.3.0~rc1/src/corelib/io/qfsfileengine_unix.cpp
--- qt4-x11-4.3.0~rc1~/src/corelib/io/qfsfileengine_unix.cpp	2007-05-06 17:54:23.000000000 +0200
+++ qt4-x11-4.3.0~rc1/src/corelib/io/qfsfileengine_unix.cpp	2007-05-17 19:10:48.000000000 +0200
@@ -830,9 +830,16 @@
     } else if (own == OwnerGroup) {
         struct group *gr = 0;
 #if !defined(QT_NO_THREAD) && defined(_POSIX_THREAD_SAFE_FUNCTIONS) && !defined(Q_OS_OPENBSD)
-        buf.resize(sysconf(_SC_GETGR_R_SIZE_MAX));
-        struct group entry;
-        getgrgid_r(ownerId(own), &entry, buf.data(), buf.size(), &gr);
+        for (unsigned size = sysconf(_SC_GETGR_R_SIZE_MAX); size < 256000; size += size)
+        {
+            buf.resize(size);
+            struct group entry;
+            // ERANGE indicates that the buffer was too small
+            if (!getgrgid_r(ownerId(own), &entry, buf.data(), buf.size(), &gr)
+                || errno != ERANGE)
+                break;
+        }
+
 #else
         gr = getgrgid(ownerId(own));
 #endif
